[Uid] Add Generate and Inspect commands

fancyweb opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

After some time using the component, I realized I often needed to quickly generate new ulids or to convert them from one format to another so I thought having those commands would be useful.

# Usage

## Generate a ULID - outputs N ULID(s) separated by new lines in base 32, base 58 or RFC 4122 format 

### Generate 1 ULID now
`php bin/console ulid:generate`

### Generate 1 ULID with a specific timestamp
`php bin/console ulid:generate --time="2021-02-02 14:00:00"`

### Generate 2 ULIDs and ouput the RFC4122 format
`php bin/console ulid:generate --count=2 --format=rfc4122`

## Generate a UUID - outputs N UUID(s) separated by new lines in RFC 4122, base 58 or base 32 format

### Generate 1 UUID (defaults from the underlying factory)
`php bin/console uuid:generate`

### Generate 1 time-based UUID now
`php bin/console uuid:generate --time-based=now`

### Generate 1 time-based UUID with a specific timestamp
`php bin/console uuid:generate --time-based="2021-02-02 14:00:00"`

### Generate 1 time-based UUID with a specific node
`php bin/console uuid:generate --time-based=now --node=fb3502dc-137e-4849-8886-ac90d07f64a7`

### Generate 1 name-based UUID (there must be a default namespace in the underlying factory)
`php bin/console uuid:generate --name-based=foo`

### Generate 1 name-based UUID with a specific namespace (overrides the default namespace from the underlying factory)
`php bin/console uuid:generate --name-based=foo --namespace=fb3502dc-137e-4849-8886-ac90d07f64a7`

### Generate 1 random-based UUID
`php bin/console uuid:generate --random-based`

### Generate 2 UUIDs and output their base 58 format
`php bin/console uuid:generate --count=2 --format=base58`

## Inspect a ULID - outputs base32, base58 and RFC 4122 formats of a ULID and its humand readable timestamp if it is time-based
`php bin/console ulid:inspect 01EWAKBCMWQ2C94EXNN60ZBS0Q`
`php bin/console ulid:inspect 1BVdfLn3ERmbjYBLCdaaLW`
`php bin/console ulid:inspect 01771535-b29c-b898-923b-b5a981f5e417`


## Inspect a UUID - outputs RFC 4122, base 58 and base 32 formats of a UUID and its human readable timestamp
`php bin/console uuid:inspect a7613e0a-5986-11eb-a861-2bf05af69e52`
`php bin/console uuid:inspect MfnmaUvvQ1h8B14vTwt6dX`
`php bin/console uuid:inspect 57C4Z0MPC627NTGR9BY1DFD7JJ`

# Register the commands

## YAML
```yaml
# services.yaml
services:
    Symfony\Component\Uid\Command\GenerateUlidCommand: ~
    Symfony\Component\Uid\Command\GenerateUuidCommand: ~
    Symfony\Component\Uid\Command\InspectUlidCommand: ~
    Symfony\Component\Uid\Command\InspectUuidCommand: ~
```

## PHP
```php
<?php

// services.php

namespace Symfony\Component\DependencyInjection\Loader\Configurator;

use Symfony\Component\Uid\Command\GenerateUlidCommand;
use Symfony\Component\Uid\Command\GenerateUuidCommand;
use Symfony\Component\Uid\Command\InspectUlidCommand;
use Symfony\Component\Uid\Command\InspectUuidCommand;

return static function (ContainerConfigurator $configurator): void {
    $services = $configurator->services()
        ->defaults()
        ->autowire()
        ->autoconfigure();

    $services
        ->set(GenerateUlidCommand::class)
        ->set(GenerateUuidCommand::class)
        ->set(InspectUlidCommand::class)
        ->set(InspectUuidCommand::class);
};
```

fancyweb committed:
[Uid] Add Generate and Inspect commands

Co-authored-by: Nicolas Grekas <nicolas.grekas@gmail.com>

stof commented:
See https://github.com/symfony/symfony/pull/36405 for previous discussion about that topic.

OskarStark commented on src/Symfony/Component/Uid/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.3.0
+-----
+
+ * added commands to generate and inspect uuid and ulid
```
```suggestion
5.3
---

 * Add commands to generate and inspect Uuid's and Ulid's
```

Fabien documented how to write `CHANGELOG.md` files here, please keep a look and adjust it accordingly.

https://github.com/symfony/symfony-docs/pull/14830

Thanks 🙏

OskarStark commented on src/Symfony/Component/Uid/composer.json:
```diff
@@ -23,6 +23,12 @@
+    "require-dev": {
+        "symfony/console": "^4.4|^5.0"
+    },
+    "suggest": {
+        "symfony/console": "For using the console commands"
+    },
```
AFAIK we don't add `suggest` anymore 🤔

fancyweb replied:
I will wait for a confirmation before removing it.

derrabus replied:
Oskar is correct, we do not add new `suggest` entries.

jderusse commented on src/Symfony/Component/Uid/Command/GenerateUlidCommand.php:
```diff
@@ -0,0 +1,62 @@
+            ->setDefinition([
+                new InputOption('count', 'c', InputOption::VALUE_OPTIONAL, 'The number of ULID to generate.', 1),
+            ])
+            ->setDescription('Generate a ULID')
```
```suggestion
            ->setDescription('Generates a ULID')
```

jderusse commented on src/Symfony/Component/Uid/Command/GenerateUlidCommand.php:
```diff
@@ -0,0 +1,62 @@
+    {
+        $this
+            ->setDefinition([
+                new InputOption('count', 'c', InputOption::VALUE_OPTIONAL, 'The number of ULID to generate.', 1),
```
Shouldn't it be `InputOption::VALUE_REQUIRED`?
The flag is optional, but when present, a value is required.

fancyweb replied:
You are right.

jderusse commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,116 @@
+                new InputOption('namespace', null, InputOption::VALUE_OPTIONAL, 'The namespace for version 3 and 5 (a UUID, in any of the supported formats (base 58, base 32 or RFC 4122)).'),
+                new InputOption('name', null, InputOption::VALUE_OPTIONAL, 'The name for version 3 and 5.'),
+            ])
+            ->setDescription('Generate a UUID')
```
generates

jderusse commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,116 @@
+        $this
+            ->setDefinition([
+                new InputArgument('version', InputArgument::OPTIONAL, 'The UUID version to generate (v1, v3, v4, v5 or v6).', 'v1'),
+                new InputOption('count', 'c', InputOption::VALUE_OPTIONAL, 'The number of UUID to generate.', 1),
```
same here, `VALUE_REQUIRED`

jderusse commented on src/Symfony/Component/Uid/Command/InspectUlidCommand.php:
```diff
@@ -0,0 +1,74 @@
+            ->setDefinition([
+                new InputArgument('ulid', InputArgument::REQUIRED, 'The ULID to inspect, in any of the supported formats (base 58, base 32 or RFC 4122)'),
+            ])
+            ->setDescription('Inspect a ULID')
```
`Inspects`

jderusse commented on src/Symfony/Component/Uid/Command/InspectUuidCommand.php:
```diff
@@ -0,0 +1,90 @@
+            ->setDefinition([
+                new InputArgument('uuid', InputArgument::REQUIRED, 'The UUID to inspect, in any of the supported formats (base 58, base 32 or RFC 4122)'),
+            ])
+            ->setDescription('Inspect a UUID')
```
`inspects`

fabpot commented:
Still :-1: for me.

nicolas-grekas commented:
The commands could be provided as a third party bundle.

Alternatively, we could provide them in the source, but not register them as commands, so that ppl could enable them by adding a line in their `services.yaml` file:

```yaml
services:
    Symfony\Component\Uid\Command\GenerateUlidCommand: ~
```

This approach could enable us to provide commands without cluttering the `list` command by default.

WDYT @fabpot?

fabpot commented:
Sure, if they are hidden by default, no problems for me.

fancyweb commented:
Unfortunately I did not know there was already another PR for that feature in the past. I still think it's an useful feature to offer. I have removed the commands registration in the FWB and fixed the other comments. I also changed the timestamp display to be in milliseconds. I still need to fix the tests.

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,116 @@
+    {
+        $this
+            ->setDefinition([
+                new InputArgument('version', InputArgument::OPTIONAL, 'The UUID version to generate (v1, v3, v4, v5 or v6).', 'v1'),
```
should it default to v4 instead?
or do we make the argument mandatory? Then, we could rename the command `uid:generate`, and make it handle ulid and uuids?

stof replied:
I would indeed make it required

nicolas-grekas replied:
@fancyweb https://github.com/symfony/symfony/pull/36405 did so, maybe you can borrow some ideas from there?

fancyweb replied:
I agree that we can make it required but I would still vote personally to separate the uuid and ulid (even if in this command it's not a requirement).

nicolas-grekas commented on src/Symfony/Component/Uid/Command/InspectUlidCommand.php:
```diff
@@ -0,0 +1,74 @@
+
+class InspectUlidCommand extends Command
+{
+    protected static $defaultName = 'ulid:inspect';
```
what about a single `uid:inspect` command that could deal with ULID and UUID?

fancyweb replied:
I would prefer to have separate commands, otherwise we need an extra argument to specify if the inspected uid is supposed to be an ulid or an uuid. We cannot guess it.

stof replied:
actually, we can: `$isUlid = \strlen($uid) === 26` if you want the strict definition.
And if you use `Ulid::fromString()`, there is no point distinguishing as you can pass a uuid to `Ulid::fromString` and it will be converted (both Ulid and Uuid are different representations of a 128bits number).

fancyweb replied:
There is a difference on the timestamp part because of different epochs. If I want to inspect `3a2be45c-5a62-11eb-9956-537bd3b4c9e7`, we cannot know if I want to interpret it as an ulid or as an uuidv1. The goal is to support any input format, not just the canonical one.

nicolas-grekas reviewed (commented):
some thougths

fancyweb commented on src/Symfony/Component/Uid/Tests/Command/InspectUuidCommandTest.php:
```diff
@@ -0,0 +1,216 @@
+  Base 58                4nwhs6vwvNU2AbcCSD1XP8                           
+  Base 32                0YPPCKFC57CA4AGREV7Q1DHN6V                       
+ ---------------------- ------------------------------------------------- 
+  Timestamp              1610977552.4279 (2021-01-18T13:45:52.428+00:00)  
```
@nicolas-grekas @jderusse `1610977552.4279` looks invalid to me, since the timestamp is supposed to be precise in milliseconds only, we should round somewhere isn'it?

fancyweb replied:
I just checked and it's actually OK.

maxhelias commented:
Is it possible to have a format option (rfc, binary..) ?

OskarStark commented on src/Symfony/Component/Uid/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 =========
 
+5.3.0
+-----
```
```suggestion
5.3
---
```

OskarStark commented on src/Symfony/Component/Uid/Command/GenerateUlidCommand.php:
```diff
@@ -0,0 +1,60 @@
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Uid\Ulid;
+
+class GenerateUlidCommand extends Command
```
Maybe you would like to add @author annotations?

fancyweb replied:
I prefer to not add them.

OskarStark replied:
Fair enough

fancyweb commented:
I pushed a new implementation that leverages the new factories.

@maxhelias I added a --format option to the generate commands.

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUlidCommand.php:
```diff
@@ -0,0 +1,118 @@
+                new InputOption('count', 'c', InputOption::VALUE_REQUIRED, 'The number of ULID to generate', 1),
+                new InputOption('format', 'f', InputOption::VALUE_REQUIRED, 'The ULID output format: canonical, base_58 or rfc_4122', 'canonical'),
+            ])
+            ->setDescription('Generates a ULID')
```
should account for https://github.com/symfony/symfony/pull/39851

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUlidCommand.php:
```diff
@@ -0,0 +1,118 @@
+            ->setDefinition([
+                new InputOption('timestamp', null, InputOption::VALUE_REQUIRED, 'The ULID timestamp: a parsable date/time string by the \DateTimeImmutable constructor. It must be greater than or equals to the UNIX epoch (1970-01-01 00:00:00)'),
+                new InputOption('count', 'c', InputOption::VALUE_REQUIRED, 'The number of ULID to generate', 1),
+                new InputOption('format', 'f', InputOption::VALUE_REQUIRED, 'The ULID output format: canonical, base_58 or rfc_4122', 'canonical'),
```
--base=16/32/58 - no need for the canonical way, as that's the default?

fancyweb replied:
In other commands with a "format" option, we include the default format in the listed possibilities.

nicolas-grekas replied:
base32, base58, rfc4122

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUlidCommand.php:
```diff
@@ -0,0 +1,119 @@
+            try {
+                $time = new \DateTimeImmutable($time);
+            } catch (\Exception $e) {
+                $io->error(sprintf('Invalid timestamp "%s". It cannot be parsed.', $time));
```
maybe reuse $e->getMessage(), after removing the `DateTimeImmutable::__construct(): ` prefix?

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUlidCommand.php:
```diff
@@ -0,0 +1,119 @@
+    {
+        $this
+            ->setDefinition([
+                new InputOption('timestamp', null, InputOption::VALUE_REQUIRED, 'The ULID timestamp: a parsable date/time string by the \DateTimeImmutable constructor. It must be greater than or equals to the UNIX epoch (1970-01-01 00:00:00)'),
```
`--time` instead, since this is a date and not a timestamp?

fancyweb replied:
Ok for the option name, but can we keep timestamp for all the descriptions since this is the name used in both RFCs?

fancyweb replied:
I changed to --time.

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+    {
+        $this
+            ->setDefinition([
+                new InputOption('time-based', null, InputOption::VALUE_REQUIRED, 'The timestamp, to generate a time based UUID: a parsable date/time string by the \DateTimeImmutable constructor. It must be greater than or equals to the UUID epoch (1582-10-15 00:00:00)'),
```
`A parsable date/time string` would be enough for the description

fancyweb replied:
Should I remove the "it must be greater ..." part too?

fancyweb replied:
I kept the last part.

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+        $this
+            ->setDefinition([
+                new InputOption('time-based', null, InputOption::VALUE_REQUIRED, 'The timestamp, to generate a time based UUID: a parsable date/time string by the \DateTimeImmutable constructor. It must be greater than or equals to the UUID epoch (1582-10-15 00:00:00)'),
+                new InputOption('node', null, InputOption::VALUE_REQUIRED, "The time based UUID's node: a UUID, in any of the supported formats (base 58, base 32 or RFC 4122)"),
```
A UUID whose node part should be used as the node of the generated UUID.

fancyweb replied:
~I drop the "in any support format" part?~ yes

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+            ->setDefinition([
+                new InputOption('time-based', null, InputOption::VALUE_REQUIRED, 'The timestamp, to generate a time based UUID: a parsable date/time string by the \DateTimeImmutable constructor. It must be greater than or equals to the UUID epoch (1582-10-15 00:00:00)'),
+                new InputOption('node', null, InputOption::VALUE_REQUIRED, "The time based UUID's node: a UUID, in any of the supported formats (base 58, base 32 or RFC 4122)"),
+                new InputOption('name-based', null, InputOption::VALUE_REQUIRED, 'The name, to generate a name based UUID'),
```
name-based UUID

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+                new InputOption('time-based', null, InputOption::VALUE_REQUIRED, 'The timestamp, to generate a time based UUID: a parsable date/time string by the \DateTimeImmutable constructor. It must be greater than or equals to the UUID epoch (1582-10-15 00:00:00)'),
+                new InputOption('node', null, InputOption::VALUE_REQUIRED, "The time based UUID's node: a UUID, in any of the supported formats (base 58, base 32 or RFC 4122)"),
+                new InputOption('name-based', null, InputOption::VALUE_REQUIRED, 'The name, to generate a name based UUID'),
+                new InputOption('namespace', null, InputOption::VALUE_REQUIRED, "The name based UUID's namespace: a UUID, in any of the supported formats (base 58, base 32 or RFC 4122)"),
```
A UUID to use at the namespace for named-based UUIDs.

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+                new InputOption('node', null, InputOption::VALUE_REQUIRED, "The time based UUID's node: a UUID, in any of the supported formats (base 58, base 32 or RFC 4122)"),
+                new InputOption('name-based', null, InputOption::VALUE_REQUIRED, 'The name, to generate a name based UUID'),
+                new InputOption('namespace', null, InputOption::VALUE_REQUIRED, "The name based UUID's namespace: a UUID, in any of the supported formats (base 58, base 32 or RFC 4122)"),
+                new InputOption('random-based', null, InputOption::VALUE_NONE, 'To generate a random based UUID'),
```
random-based

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+                    try {
+                        $node = Uuid::fromString($node);
+                    } catch (\InvalidArgumentException $e) {
+                        $io->error(sprintf('Invalid node "%s". It cannot be parsed.', $node));
```
let's reuse $e->getMessage as much as possible (same everywhere)

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+                    return 2;
+                }
+
+                if ($time < new \DateTimeImmutable('@-12219292800')) {
```
we don't have to check for these, the implementations do it already

fancyweb replied:
I thought it would be better to check it to display a better error message. It is better than an uncaught exception + "The given UUID date cannot be earlier than 1582-10-15" IMO.

nicolas-grekas replied:
We can catch the logic exception and use $e->getMessage() instead.
We should not duplicate such logic in commands IMHO

nicolas-grekas commented on src/Symfony/Component/Uid/Command/InspectUlidCommand.php:
```diff
@@ -0,0 +1,75 @@
+    {
+        $this
+            ->setDefinition([
+                new InputArgument('ulid', InputArgument::REQUIRED, 'The ULID to inspect, in any of the supported formats (base 58, base 32 or RFC 4122)'),
```
No need for the part about formats imho

nicolas-grekas commented on src/Symfony/Component/Uid/Command/InspectUlidCommand.php:
```diff
@@ -0,0 +1,75 @@
+            ])
+            ->setDescription(self::$defaultDescription)
+            ->setHelp(<<<'EOF'
+The <info>%command.name%</info> display information about a ULID.
```
displays

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+
+    <info>php %command.full_name%</info>
+
+To generate a time based UUID:
```
time-based (same everywhere, please add the dash)

nicolas-grekas commented on src/Symfony/Component/Uid/Command/InspectUuidCommand.php:
```diff
@@ -0,0 +1,89 @@
+            ])
+            ->setDescription(self::$defaultDescription)
+            ->setHelp(<<<'EOF'
+The <info>%command.name%</info> display information about a UUID.
```
displays

nicolas-grekas commented on src/Symfony/Component/Uid/Command/InspectUlidCommand.php:
```diff
@@ -0,0 +1,75 @@
+        $io = new SymfonyStyle($input, $output instanceof ConsoleOutputInterface ? $output->getErrorOutput() : $output);
+
+        try {
+            /** @var Ulid $ulid */
```
this should not be needed since the docblock says `@return static`

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,194 @@
+                        return 4;
+                    }
+                } else {
+                    $refl = new \ReflectionProperty($this->factory, 'nameBasedNamespace');
```
not sure why we'd need this - the factory already checks that
can't we do without?

fancyweb replied:
The factory throws a \LogicException "A namespace should be defined when using UuidFactory::nameBased()". As an average user, when I use the command, I probably do not know about the factory, if I see this error message it is not going to help me a lot.

ro0NL commented:
> Sure, if they are hidden by default, no problems for me.

arent we registering them as hidden? being able to default advertise `bin/console ..` seems reasonable

ro0NL commented on src/Symfony/Component/Serializer/Normalizer/UidNormalizer.php:
```diff
@@ -22,9 +22,9 @@ final class UidNormalizer implements NormalizerInterface, DenormalizerInterface,
-    public const NORMALIZATION_FORMAT_RFC_4122 = 'rfc_4122';
+    public const NORMALIZATION_FORMAT_BASE58 = 'base58';
+    public const NORMALIZATION_FORMAT_BASE32 = 'base32';
+    public const NORMALIZATION_FORMAT_RFC4122 = 'rfc4122';
```
i feel like we should move the consts to `AbstractUid`

nicolas-grekas replied:
Why? `AbstractUid` nor the `Uid` component wouldn't use them anywhere.

ro0NL replied:
because the formats are leaking now: `The ULID output format: base32, base58 or rfc4122`

nicolas-grekas replied:
I think these consts in Serializer are legit.
Then, the duplication in Uid is legit also - these are from different domains.
Also, the commands (on the cli), use inline values.
In the end, these consts in Uid won't serve any purpose for the devs.
They would actually make the public API of the component more complex to me.
I'd prefer not to add them as you might have understood :)

ro0NL replied:
i think both serializer+console are duplicating what could be `AbstractUid::toFormat($format)`

nicolas-grekas replied:
Oh, then I'd recommend splitting this discussion into a separate PR or issue.

nicolas-grekas commented:
> arent we registering them as hidden?

Nope: we're *not registering* them by default, so that we'll ask ppl to opt-in to list them.

nicolas-grekas commented on src/Symfony/Component/Uid/Command/GenerateUuidCommand.php:
```diff
@@ -0,0 +1,200 @@
+                };
+                break;
+
+            case false !== $random:
```
`case $random:`

nicolas-grekas reviewed (approved):
Don't miss updating the description of the PR :)

nicolas-grekas commented:
Thank you @fancyweb.
